fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulktools/Main.java

```diff
@@ -175,50 +223,81 @@ private int getFhirResourceNumberFromBufferReader(BufferedReader resReader) thro
 
 
 
-    private void writeFhirResourceFromBufferReader(BufferedReader resReader,  AmazonS3 cosClient, ByteArrayOutputStream bufferStream, int numOfRes4Seg) throws Exception {
+    private boolean writeFhirResourceFromBufferReader(BufferedReader resReader,  AmazonS3 cosClient, ByteArrayOutputStream bufferStream, long num4Seg, boolean isSegBySize) throws Exception {
         int lineRed = 0;
         int segNum = 0;
+        long segSize = 0;
+        HashSet<String> UniqueResourceIds= new HashSet<String>();
+
         List<PartETag> dataPackTags = new ArrayList<>();
         String uploadId = null;
         int partNum = 1;
         boolean isMore2Read = true;
+        boolean isAbortStream = false;
+        int totalRead = 0;
         while (isMore2Read) {
-                String resLine = resReader.readLine();
-                lineRed++;
-                if (resLine == null) {
-                    isMore2Read = false;
-                } else {
+            String resLine = resReader.readLine();
+            lineRed++;
+            totalRead++;
+            if (resLine == null) {
+                isMore2Read = false;
+            } else {
+                boolean isToAdd = true;
+                if (!isSegBySize && numberOfTotalResources > 1000) {
+                    try {
+                        Resource res = FHIRParser.parser(Format.JSON).parse(new StringReader(resLine));
+                        isToAdd = UniqueResourceIds.add(res.getId());
+                    } catch (FHIRParserException ex) {
+                        isToAdd = false;
+                    }
+                }
+
+                if (isToAdd) {
                     bufferStream.write(resLine.getBytes());
                     bufferStream.write(Constants.NDJSON_LINESEPERATOR);
+                    segSize += (resLine.getBytes().length + Constants.NDJSON_LINESEPERATOR.length);
+                } else {
+                    lineRed--;
                 }
+            }
 
-                if (bufferStream.size() > Constants.COS_PART_MINIMALSIZE
-                        || (segNum < numberOfFiles - 1 && lineRed == numOfRes4Seg)
-                        || !isMore2Read) {
-                    String segName = cosFile2Break + "_seg" + segNum;
-                    if (uploadId == null) {
-                        uploadId = BulkDataUtils.startPartUpload(cosClient, cosBucketName, segName, true);
-                    }
+            if (bufferStream.size() > Constants.COS_PART_MINIMALSIZE
```

Comment:
If you are splitting resources into separate files, how do you ensure that the same resource or ID is in the same file?

Reply:
the tool doesn't really keep all the duplicated resources, it only keep one of them for now.
e.g, in Organization, they are more than 2K lines with the exactly the same contents, the tool will only keep one of them.

Reply:
right - it's not a duplicate resource though... it's a historic resource. 

Is the tool removing the historic resource? e.g. 
if there are two versions, it'll only insert version 2?

Reply:
In the organization data set, there is no version info in the resource, they are simply duplicated lines of the same contents ... the tool simply keep the first one it runs into and drops all the others.  I thought about making the tool better/smarter to be able to move all duplicated/historic resources into the same file, but because the data set could be huge which could cause huge memory consumption to track which Id is store in which file, so this is why I just made the tool simply for now drop the others resources with the same ids ...

Reply:
but It's not bad idea to enhance this tool in the future to make it a real fhir data per-processing tool, e.g, not only put the resources with the same id into the same file, but also can validate the resource etc ...

Reply:
Is this tool only used in testing?  perhaps move it to src/test/ 
not the src/main

Reply:
the tool itself is not for any test purpose, I used it mainly as an utility tool to break the huge one piece COS file into multiple to allow them to be able to processed by multiple partitions in parallel ...